File: frontend/src/utils/exceptions.js
/**
 * React port of utils/exceptions.py's custom exception hierarchy.
 * Used by the mocked API services / pages to model the same error
 * categories the Streamlit app switched on to pick an error message
 * and icon.
 */
export class VideoAssistantError extends Error {
  constructor(message) {
    super(message);
    this.name = "VideoAssistantError";
  }
}

export class InvalidYouTubeURLError extends VideoAssistantError {
  constructor(message) {
    super(message);
    this.name = "InvalidYouTubeURLError";
  }
}

export class TranscriptNotFoundError extends VideoAssistantError {
  constructor(message) {
    super(message);
    this.name = "TranscriptNotFoundError";
  }
}

export class TranscriptFetchError extends VideoAssistantError {
  constructor(message) {
    super(message);
    this.name = "TranscriptFetchError";
  }
}

export class EmbeddingGenerationError extends VideoAssistantError {
  constructor(message) {
    super(message);
    this.name = "EmbeddingGenerationError";
  }
}

export class VectorStoreError extends VideoAssistantError {
  constructor(message) {
    super(message);
    this.name = "VectorStoreError";
  }
}

export class LLMGenerationError extends VideoAssistantError {
  constructor(message) {
    super(message);
    this.name = "LLMGenerationError";
  }
}

export class EmptyQuestionError extends VideoAssistantError {
  constructor(message) {
    super(message);
    this.name = "EmptyQuestionError";
  }
}

File: backend/utils/exceptions.py
"""
utils/exceptions.py

Custom exception classes used throughout the application to provide
clear, specific error handling and user-friendly messages.

Reused verbatim from the original Streamlit project's
utils/exceptions.py — no behavior, names, or messages changed.
"""


class VideoAssistantError(Exception):
    """Base exception for all application-specific errors."""
    pass


class InvalidYouTubeURLError(VideoAssistantError):
    """Raised when the provided YouTube URL is invalid or video ID cannot be extracted."""
    pass


class TranscriptNotFoundError(VideoAssistantError):
    """Raised when a transcript is unavailable, disabled, or missing for a video."""
    pass


class TranscriptFetchError(VideoAssistantError):
    """Raised when the transcript fetch operation fails for any other reason."""
    pass


class EmbeddingGenerationError(VideoAssistantError):
    """Raised when embedding generation fails."""
    pass


class VectorStoreError(VideoAssistantError):
    """Raised when a ChromaDB operation fails (insertion, query, persistence)."""
    pass


class LLMGenerationError(VideoAssistantError):
    """Raised when the LLM API call fails or returns an invalid response."""
    pass


class EmptyQuestionError(VideoAssistantError):
    """Raised when the user submits an empty or whitespace-only question."""
    pass
